File: c++/Spiral matrix.cpp
Spiral matrix on leetcode:- https://leetcode.com/problems/spiral-matrix/ 

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ans;
        int row= matrix.size(), col= matrix[0].size();
        int left=0, right= col-1, top=0, bottom= row-1;

        while(left<= right && top<=bottom)
        {
            for(int i=left; i<=right; i++)
            {
                ans.push_back(matrix[top][i]);
            }
            top++;
            for(int i=top; i<=bottom; i++)
            {
                ans.push_back(matrix[i][right]);
            }
            right--;
            if(top<=bottom)
            {
            for(int i=right; i>=left; i--)
            {
                ans.push_back(matrix[bottom][i]);
            }
            bottom--;
            }
            if(left<=right)
            {
            for(int i=bottom; i>=top; i--)
            {
                ans.push_back(matrix[i][left]);
            }
            left++;
            }
        }
        return ans;
    }
};

File: JAVA/spiral_matrix.java
// https://leetcode.com/problems/spiral-matrix/description/

// Given an m x n matrix, return all elements of the matrix in spiral order.

 

// Example 1:


// Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
// Output: [1,2,3,6,9,8,7,4,5]
// Example 2:


// Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
// Output: [1,2,3,4,8,12,11,10,9,5,6,7]
 

// Constraints:

// m == matrix.length
// n == matrix[i].length
// 1 <= m, n <= 10
// -100 <= matrix[i][j] <= 100

import java.util.ArrayList;
import java.util.List;

public class spiral_matrix {
    public List<Integer> spiralOrder(int[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;

        List<Integer> ans = new ArrayList<>();

        int top = 0, left = 0, right = m - 1, bottom = n-1;

        while(top <= bottom && left <= right){
            for(int i = left; i <= right; i++){
                 ans.add(matrix[top][i]);
            }
            top++;
            for(int i = top; i <= bottom; i++){
                ans.add(matrix[i][right]);
            }
            right--;
            if(top <= bottom){
                for(int i = right; i >= left; i--){
                    ans.add(matrix[bottom][i]);
                }
                bottom--;
            }
            if(left <= right){
                for(int i = bottom; i >= top; i--){
                    ans.add(matrix[i][left]);
                }
                left++;
            }
        }
        return ans;
    }
}
